beefy-gadget/src/worker.rs

```diff
@@ -402,3 +397,46 @@ where
 
 	header.digest().convert_first(|l| l.try_to(id).and_then(filter))
 }
+
+#[cfg(test)]
+mod tests {
+	use sp_runtime::SaturatedConversion;
+
+	// This replicates the `should_vote_on()` core algorithm.
+	// We use that as an interim solution, until we can test `BeefyWorker` proper
+	fn should_vote_on(number: u64, best_grandpa: u64, best_beefy: u64, min_delta: u64) -> bool {
+		let diff = best_grandpa.saturating_sub(best_beefy);
+		let diff = diff.saturated_into::<u32>();
+		let next_power_of_two = (diff / 2).next_power_of_two() as u64;
+		let next_block_to_vote_on = best_beefy + min_delta.max(next_power_of_two);
+		number == next_block_to_vote_on
+	}
```

Comment:
No, duplicating the code is not the way to go.

Please move this function one scope up (simply to the `worker` module) and change `BeefyWorker::should_vote_on` to simply call this function. Rough implementation:
```
impl BeefyWorker {
...
  fn should_vote_on(&self, number: NumberFor<B>) -> bool {
    should_vote_on(number, self.best_grandpa_block, self.best_beefy_block, self.min_block_delta)
  }
...
}
```

Reply:
Damn, should have read the book ;-)